samples/21.core-bot-app-insights/src/main/java/com/microsoft/bot/sample/corebot/app/insights/AdapterWithErrorHandler.java

```diff
@@ -0,0 +1,111 @@
+     *
+     * <p>
+     * For this sample, a simple message is displayed. For a production Bot, a more
+     * informative message or action is likely preferred.
+     * </p>
+     *
+     * @param withConfiguration The Configuration object to use.
+     */
+    public AdapterWithErrorHandler(
+        Configuration withConfiguration,
+        TelemetryInitializerMiddleware telemetryInitializerMiddleware,
+        BotTelemetryClient botTelemetryClient,
+        @Nullable ConversationState withConversationState) {
+        super(withConfiguration);
+        this.use(telemetryInitializerMiddleware);
+
+        // Use telemetry client so that we can trace exceptions into Application Insights
+        this.adapterBotTelemetryClient = botTelemetryClient;
+        setOnTurnError((turnContext, exception) -> {
+            // Track exceptions into Application Insights
+            // Set up some properties for our exception tracing to give more information
+            Map<String, String> properties = new HashMap<String, String>();
+            properties.put("Bot exception caught in", "AdapterWithErrorHandler - OnTurnError");
+
+            // Send the exception telemetry:
+            adapterBotTelemetryClient.trackException((Exception) exception, properties, null);
+
+            // Log any leaked exception from the application.
+            // NOTE: In production environment, you should consider logging this to
+            // Azure Application Insights. Visit https://aka.ms/bottelemetry to see how
+            // to add telemetry capture to your bot.
+            LoggerFactory.getLogger(AdapterWithErrorHandler.class).error("onTurnError", exception);
+
+            // Send a message to the user
+            return turnContext.sendActivities(
+                MessageFactory.text(ERROR_MSG_ONE), MessageFactory.text(ERROR_MSG_TWO)
+            ).thenCompose(resourceResponse -> sendTraceActivity(turnContext, exception))
+                .thenCompose(stageResult -> {
+                    if (withConversationState != null) {
+                        // Delete the conversationState for the current conversation to prevent the
+                        // bot from getting stuck in a error-loop caused by being in a bad state.
+                        // ConversationState should be thought of as similar to "cookie-state" in a
+                        // Web pages.
+                        return withConversationState.delete(turnContext)
+                            .exceptionally(deleteException -> {
+                                LoggerFactory.getLogger(AdapterWithErrorHandler.class)
+                                    .error("ConversationState.delete", deleteException);
+                                return null;
+                            });
+                    }
+                    return CompletableFuture.completedFuture(null);
+                });
+        });
+    }
+
+    private CompletableFuture<Void> sendTraceActivity(
```

Comment:
not sure if we need sendTraceActivity as a separate method in this case. The class in i[ntegration](https://github.com/southworks/botbuilder-java/blob/internal/feature/southworks/samples/corebot-app-insights/adapterwitherrorhandler/libraries/bot-integration-core/src/main/java/com/microsoft/bot/integration/AdapterWithErrorHandler.java#L92) has it because it's called from both constructors. Nevermind anyway, does no harm either

Reply:
I understand what you say, I had the same doubt, however, it's just a copy paste from the main [AdapterWithErrorHandler](https://github.com/microsoft/botbuilder-java/blob/main/libraries/bot-integration-core/src/main/java/com/microsoft/bot/integration/AdapterWithErrorHandler.java)

Reply:
precisely, that's the case I mentioned. C# doesn't have that. It's ok anyway